qiskit/compiler/assemble.py

```diff
@@ -16,128 +16,136 @@
 import uuid
 import copy
 
-from qiskit.circuit import QuantumCircuit
+from typing import Union, List, Dict, Optional
+from qiskit.circuit import QuantumCircuit, Qubit
 from qiskit.exceptions import QiskitError
 from qiskit.pulse import ScheduleComponent, LoConfig
 from qiskit.assembler.run_config import RunConfig
 from qiskit.assembler import assemble_circuits, assemble_schedules
-from qiskit.qobj import QobjHeader
+from qiskit.qobj import QobjHeader, Qobj
 from qiskit.qobj.utils import MeasLevel, MeasReturnType
 from qiskit.validation.jsonschema import SchemaValidationError
+from qiskit.providers import BaseBackend
+from qiskit.pulse.channels import PulseChannel
 
 
 # TODO: parallelize over the experiments (serialize each separately, then add global header/config)
-def assemble(experiments,
-             backend=None,
-             qobj_id=None, qobj_header=None,
-             shots=None, memory=False, max_credits=None, seed_simulator=None,
-             qubit_lo_freq=None, meas_lo_freq=None,
-             qubit_lo_range=None, meas_lo_range=None,
-             schedule_los=None, meas_level=MeasLevel.CLASSIFIED,
-             meas_return=MeasReturnType.AVERAGE, meas_map=None,
-             memory_slot_size=100, rep_time=None, parameter_binds=None,
-             parametric_pulses=None,
-             **run_config):
+def assemble(experiments: Union[QuantumCircuit, List[QuantumCircuit]],
+             backend: Optional[BaseBackend] = None,
+             qobj_id: Optional[str] = None, qobj_header: Optional[Union[QobjHeader, Dict]] = None,
+             shots: Optional[int] = None, memory: Optional[bool] = False,
+             max_credits: Optional[int] = None,
+             seed_simulator: Optional[int] = None,
+             qubit_lo_freq: Optional[List[int]] = None, meas_lo_freq: Optional[List[int]] = None,
+             qubit_lo_range: Optional[List[int]] = None, meas_lo_range: Optional[List[int]] = None,
+             schedule_los: Optional[Union[List[Union[Dict[PulseChannel, float], LoConfig]],
```

Comment:
I don't know if we want to do this or not, but types can be aliased:
```
# or ScheduleLoOptions for consistency for now
ScheduleFrequencyOptions = Optional[Union[List[Union[Dict[PulseChannel, float], LoConfig]]...
```
and then in the function you'd just have to write `schedule_los: ScheduleFrequencyOptions`

Reply:
This could also be nice for:
```
ExperimentType: Union[Circuit, List[Circuit], Schedule, List[Schedule]]
```

Reply:
In the first example you give, where would the `ScheduleFrequencyOptions` definition go? Would that just be at the top of the file similar to imports?

Reply:
yes, exactly, right after the imports. You can document it like a module variable:
```
Name = Union[etc]
"""Documentation string."""
```

Reply:
This would look nicer considering it is a very long type hint. Is this something we want to do? If so I can add it in.

Reply:
Is there some other meaningful chunks that this can be broken down into (and then those chunks aliased)? aliasing is good to help humans parse what's going on, but in this case just aliasing the whole thing still leaves the alias hard to parse :/